src/pudl/io_managers.py

```diff
@@ -568,12 +568,17 @@ def __init__(self, base_path: UPath, schema: pa.Schema) -> None:
         super().__init__(base_path=base_path)
         self.schema = schema
 
-    def dump_to_path(self, context: OutputContext, obj: pd.DataFrame, path: UPath):
+    def dump_to_path(self, context: OutputContext, obj: dd.DataFrame, path: UPath):
         """Write dataframe to parquet file."""
-        raise NotImplementedError("This IO Manager doesn't support writing data.")
+        logger.info(f"Write df to parquet at {path}")
+        schema = Resource.from_id("hourly_emissions_epacems").to_pyarrow()
+        obj.repartition(npartitions=1).to_parquet(
```

Comment:
Ok I've played around with this a bunch, and I have a solution that is *mostly* working as expected. First, `load_from_path` is returning a Dask dataframe, not pandas dataframe, so by default `to_parquet` is going to create a partitioned parquet directory. Using `repartition(npartitions=1)` before writing will create one monolithic file. The one slight oddity is that it still creates a directory with the single file in it: `epacems_output/part.0.parquet`.

I also tried a different solution reading each partition into memory, then appending to the same file. This worked exactly as expected, but it was really slow (perhaps this could be fixed by using larger partitions like just `year` parts instead of `year`/`state` combos. Here's the code for that:

```py
with pa.parquet.ParquetWriter(where=str(path), schema=shcema, compression="snappy") as pqwriter:
    for part in EpaCemsSettings().partitions:
        year = part["year"]
        state = part["state"]
        
        df = obj[(obj.year == year) & (obj.state == state)].compute()
        pqwriter.write_table(
            pa.Table.from_pandas(df, schema=schema, preserve_index=False)
        )
```

All of this being said, while working on this I've been wondering if it would just be together to zip together all of the partitioned parquet files. This would allow to access all the data with a single download, while still being able to use the partitioning. Curious what @zaneselvans @bendnorman and @jdangerx think would be the best solution here?

Reply:
There's a lot of functionality for reading parquet file metadata remotely out of object stores, and interacting with only the desired columns and row-groups, and that functionality would go away if the parquet files were zipped into a single file.

Any one year of data should easily fit into memory so doing one year at a time seems like a good thing to try. You might need to do something to ensure that both year and state row-groups are preserved in a useful way in the monolithic output though.

Reply:
Another option is to not use an IO Manager and just write the partitions to a monolithic file when they are created in `pudl.etl.epacems_assets.hourly_emissions_epacems()`. @zschira you could reuse this code from `main`:

https://github.com/catalyst-cooperative/pudl/blob/73eb0bad6e6154b17da2166353a839d85cc142d7/src/pudl/etl.py#L299-L324